### ISSUE TITLE
rowHeight not refreshing while dragging

### ISSUE THREAD

**Issue by svenadlung** [fritz-c/react-sortable-tree#264]
We are working with a `rowHeight`-function to give the nodes different heights depending on the nested level of each node. This is working well so far. 

But we ran into problems while dragging a nodes. The calculated heights need to be recalculated while dragging, but they don't. We noticed this is working well with `isVirtualized` is false, but not if it's set to true.

Anyone got an idea?

**Comment by svenadlung** [comment]
To be more precise I think height is not re-set at all until toggling visibilities or removing and adding nodes. Even updating the state (changing sorting) is causing the issue. 

Can anybody provide help please?

**Comment by jcalebj** [comment]
I came across the same problem as @svenadlung 
Any ideas?

**Comment by asynchub** [comment]
Nice component:) and here is the same thing about rowHeight being invoked only at onVisibilityToggle, if isVirtualized={true}. @checklist was going to resolve this via custom NodeRendererDefault in issue #192  Does not seem to be shortcut solution. Would be cool to invoke rowHeight programmatically.

**Comment by jcalebj** [comment]
So what if at this [line](https://github.com/fritz-c/react-sortable-tree/blob/master/src/react-sortable-tree.js#L666) there was: 
```
style: {
            ...rowStyle,
            height:
              typeof rowHeight !== 'function'
                ? rowHeight
                : rowHeight({
                    index,
                    treeIndex: index,
                    node: row.node,
                    path: row.path,
                  }),
          },

**Comment by asynchub** [comment]
nice hack @jcalebj 👍 
instead of row.node and row.path apply rows[index].node and rows[index].path there.
rowHeight invoked now, node height is changed, but next nodes are "not moved down" when node height button is toggled. setState does not affect other rows movement either before or after.
experimented with transpiled code:
style: Object.assign({}, rowStyle, {
                                          height: typeof rowHeight !== 'function' ? rowHeight : rowHeight({
                                            index: index,
                                            treeIndex: index,
                                            node: rows[index].node,
                                            path: rows[index].path
                                          })
                                        }),

**Comment by asynchub** [comment]
to [ScrollZoneVirtualList](https://github.com/fritz-c/react-sortable-tree/blob/7c842aaf1afa0ab106164f5000c3cea0865cea82/src/react-sortable-tree.js#L635), which is actually `withScrolling from 'react-dnd-scrollzone'` there lines like theese are passed:
[              height={height}](https://github.com/fritz-c/react-sortable-tree/blob/7c842aaf1afa0ab106164f5000c3cea0865cea82/src/react-sortable-tree.js#L646)
[              estimatedRowSize={
                typeof rowHeight !== 'function' ? rowHeight : undefined
              }](https://github.com/fritz-c/react-sortable-tree/blob/7c842aaf1afa0ab106164f5000c3cea0865cea82/src/react-sortable-tree.js#L649)
interesting, if they can affect another rows to move down on toggling the target row?
need to check that later today.

**Comment by jcalebj** [comment]
@asynchub I spent the majority of the day yesterday trying to track down the reason for this behavior 😒 
I traced it down through `ScrollZoneVirtualList` to `react-virtualized` and found this [issue](https://github.com/bvaughn/react-virtualized/issues/23) this morning.  It seems that the second to last comment may be the solution we've been looking for 😻 I'll report back with my results

**Comment by asynchub** [comment]
thanks a lot for your find @jcalebj 👍 
having same feeling to deal with react-virtualized
hopefully you are right, that recomputeRowHeight can resolve this. 
will try this later in the evening as well :)

**Comment by svenadlung** [comment]
Yeah, I was thinking the issue is with `react-virtualized`. I didn't dig deeper so far. Did you progress your work?

**Comment by jcalebj** [comment]
Yeah @asynchub @svenadlung , I found a solution for resetting the height correctly _after_ moving/dragging a node. To be clear, the issue is **not** with `react-virtualized`.


`react-sortable-tree.js` is missing a call to `List.recomputeRowHeights()`.  See docs [here](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#recomputerowheights-index-number).

Right now I have that call within `moveNode` function (which happens after drag is finished) which updates the row heights correctly after a move happens. 

The rowHeights update while dragging is still not working though.  I don't know where to call the recomputeHeight function to update the rowHeights while dragging.  I've tried several different places in the lifecylce, but there's a point between a row of height X and a row of height Y that it sits between "slideRegion" of _both_ rows, so it shifts back and forth from below to above the hovered row.  You can see this behavior in the attached gif.  I have a debounce in place to slow it down, else it happens so often you can't see it and it nearly crashes.  You can see that if I leave that particular "middle ground" slide region and go lower or higher than the row, that it resolves and sets the correct rowHeight both when it's above and below the row with the larger rowHeight.

![rowheight issue](https://user-images.githubusercontent.com/5686190/37171828-24b4223e-22dd-11e8-99f3-ff17ec07a17a.gif)



I can open up a PR on what I have that fixes the rowHeight issue _after_ moving a node, but will need some more time and/or assistance resolving the problem while dragging (detailed above).  What do you guys think?

**Comment by asynchub** [comment]
hi @svenadlung have found how to invoke the function `recomputeRowHeights` inside `ScrollZoneVirtualList`.
need to find out, where to call this function exactly to make clicked node height and offset position of next rows. working on this.
any ideas @svenadlung, @jcalebj

**Comment by jcalebj** [comment]
@asynchub see my last comment for my findings

**Comment by asynchub** [comment]
@jcalebj, have you seen this plunker [https://plnkr.co/edit/lTFmKOQlJLIehL1LIY47?p=preview](url) ?
got it from 4th post by @bvaughn from this issue: [https://github.com/bvaughn/react-virtualized/issues/307](url)
there is one more `recomputeRowHeights` and that is passed to <VirtualScroll> with comment: `// Check if any heights have been recalculated before the ref was set`
have you tried to call recomputeRowHeights from both: `<ScrollZoneVirtualList>` and from it's `rowRenderer` - 'renderRow' props? `renderRow` is actually rst's prop, which you can call from component, that you return by `renderRow` function.

**Comment by jcalebj** [comment]
@asynchub I'll test that out today

**Comment by asynchub** [comment]
@jcalebj, thanks a lot for your fix
from there, got working solution for challenge with row height on tree node click:
ref line is the same and at the same place as you have:
`ref={el => {
                this.scrollZoneVirtualListComponent = el
              }}`
and the `recomputeRowHeights` call is just before you return from render, like this:
`if (isVirtualized) {
      this.scrollZoneVirtualListComponent && 
      this.scrollZoneVirtualListComponent.wrappedInstance.recomputeRowHeights();
    }`
and of course, style in `rowRenderer` at `if (isVirtualized)`:
`style: Object.assign({}, rowStyle, {
                    height: typeof rowHeight !== 'function' ? rowHeight : rowHeight({
                      index: index,
                      treeIndex: index,
                      node: rows[index].node,
                      path: rows[index].path
                    })
                  }),`
from here, it is the point attempt passing these things across components of both `react-sortable-tree` and `react-virtualized`
by the way, have you noticed?:, that `react-sortable-tree` has a good prop `reactVirtualizedListProps`, worth of trying to pass things to the List / Grid of `react-virtualized`
how is your challenge with moveNode so far? 👍

**Comment by jcalebj** [comment]
@asynchub I believe if you call `recomputeRowHeights` within the render that that's an anti-pattern. Are you not getting warnings in your console from that?

To be clear, did you get a working solution in place? Or is what you wrote above a work in progress?

I've solved the problem of having incorrect row heights after a move, it's in [PR 269](https://github.com/fritz-c/react-sortable-tree/pull/269).  The remaining problem is to get the rowHeights to updated correctly _while dragging/moving_.  Have you had any progress on that front?

**Comment by jcalebj** [comment]
@wuweiweiwu Is there a reason for labelling this issue 'enhancement' and not 'bug'?  This is a known problem with current documented functionality. Re-sorting a node when `isVirtualized` and `rowHeight` set to a func does _not_ work correctly.

**Comment by asynchub** [comment]
@jcalebj, there is warning, yes this gives a warning about `rener()` purity:
`Warning: Cannot update during an existing state transition (such as within `render` or another component's constructor).`
this is just working solution in place, just for now, and it works for dragging/moving feature too.

**Comment by svenadlung** [comment]
Any news on this? In my understanding there is a PR which is fixing this on drop but not while dragging. Anyways, the PR is not ready to merge? Please correct me if I am wrong! It would be great if someone finds a solution even while dragging (besides the hack to put it in the render-Function ;)).

**Comment by wuweiweiwu** [comment]
@svenadlung seems that @jcalebj is working on a PR right now. But please feel free to open a PR as well :)

**Comment by developerpoidi** [comment]
Are there any updates regarding this issue? 
I am calculating the height based on the text so I can have multiple lines, but with this bug my list is not really usable.

**Comment by developerpoidi** [comment]
There also seams to be a problem with the height recalculation if you move an item between a parent-child relation (where the blue arrow is shown and moves the item below it).

The problem occurs also after using the change proposed in https://github.com/frontend-collective/react-sortable-tree/pull/269 (which at least fixes the calculation after the move)

**Comment by dcporter44** [comment]
i hacked around this by forcing calling recomputeRowHeights on drag change. On drag I set an interval at 250ms to recompute row heights. When stopped dragging, I clear the interval. I originally didnt use an interval and called recomputeRowHeights within the `onChange` prop, but that caused me to call `recomputeRowHeights` hundreds of times (for every pixel i moved my cursor while dragging). Feel free to change 250ms to a lower number. But calling `recomputeRowHeights` too often will hurt performance.

`<SortableTree>
...
reactVirtualizedListProps={{
   autoHeight: true,
   ref: ref => this.list = ref;
}}
onDragStateChanged={({ isDragging }) => {
     if (this.list) {
       const recompute = () => {
          this.list.wrappedInstance.recomputeRowHeights();
        };

        if (isDragging) {
          this.dragInterval = setInterval(recompute, 250);
        } else {
          clearInterval(this.dragInterval);
          this.list.wrappedInstance.recomputeRowHeights();
        }
    }
}}
</SortableTree>`

**Comment by dilvane** [comment]
> i hacked around this by forcing calling recomputeRowHeights on drag change. On drag I set an interval at 250ms to recompute row heights. When stopped dragging, I clear the interval. I originally didnt use an interval and called recomputeRowHeights within the `onChange` prop, but that caused me to call `recomputeRowHeights` hundreds of times (for every pixel i moved my cursor while dragging). Feel free to change 250ms to a lower number. But calling `recomputeRowHeights` too often will hurt performance. This method isnt perfect, but it does the job:
> 
> ```
> <SortableTree>
> ...
> reactVirtualizedListProps={{
>    autoHeight: true,
>    ref: ref => this.list = ref;
> }}
> onDragStateChanged={({ isDragging }) => {
>      if (this.list) {
>        const recompute = () => {
>           this.list.wrappedInstance.recomputeRowHeights();
>         };
>         if (isDragging) {
>           this.dragInterval = setInterval(recompute, 250);
>         } else {
>           clearInterval(this.dragInterval);
>           this.list.wrappedInstance.recomputeRowHeights();
>         }
>     }
> }}
> </SortableTree>
> ```

Works fine @dcporter44, I had to use `this.list.wrappedInstance.current.recomputeRowHeights();` instead, probably some react-virtualized update. Thanks.

**Comment by mx2323** [comment]
i was able to get the correct behavior on my own local fork of react-sortable-tree. the changes probably are not usable, nor are they the 'right' way to do things, but i'm happy enough with the resulting behavior that it's usable for us for now. 

the issue is that if you modify the code to call recomputeRowHeights() more frequently like everyone is doing above, a larger and smaller row will end up toggling in the tree infinitely while you hover over a fixed location. by recompute row heights every time the state was changed on hover() in react-sortable-tree.js, it just  exposes a problem where moving row heights around with the current logic can cause infinite toggling of the draggedNode location. 

for instance, this happens whenever a smaller height node A is hovered over the bottom portion of a larger drop target node B, A is inserted in the tree above B in the library, but that causes the code to keep rerendering as it tries to moves A to be underneath B.  again, this only happens if you start callign recomputeRowHeights() on dragging, which is what everyone in the thread above is doing. 

i made a change to the library in dnd-manager.js so that even if it is hovering a lower portion over a component, it will recommend inserting it after instead of always inserting before the drop target. 

if people are interested, i can work on a diff. it's probably not the the _right_ way to do this, but appears to be working good enough for our purposes.

**Comment by svenadlung** [comment]
> i was able to get the correct behavior for refreshing row height while dragging on my own local fork of react-sortable-tree. the changes probably are not usable by anyone else, nor are they the 'right' way to do things, but i'm happy enough with the resulting behavior that it's usable for us for now.
> 
> the issue is that if you modify the code to call recomputeRowHeights() more frequently like everyone is doing above, a larger and smaller row will end up toggling in the tree infinitely while you hover over a fixed location with your mouse pointer. by calling recompute row heights every time the state is changed on hover() in react-sortable-tree.js, it just exposes a problem in the library where moving variable height rows around with the current logic can cause infinite toggling of the draggedNode location.
> 
> for instance, this happens whenever a smaller height node A is hovered over the bottom portion of a larger drop target node B, A is inserted in the tree above B by the library, but on redraw, if the currently hovering node A is over B, it will try to move A down and B up. But, then on the next redraw, it will detect the node is over B, and then will try to again insert A above B, which leaves you in that infinite loop of redrawing because you are recomputing row heights
> 
> i made a change to the library in dnd-manager.js so that if it is hovering over a lower portion over a component, it will try to insert it after instead of always trying to place it above the target node.
> 
> if people are interested, i can work on a diff. it's probably not the the _right_ way to do this, but appears to be working good enough for our purposes.

Exactly. We were facing the same problem, thats why we ended up only using `onMoveNode` and recompute heights only after instead of doing it while dragging.

It would be amazing to see your sources. Wasn't able to find your fork :(

**Comment by ZirkleTsing** [comment]
> Yeah @asynchub @svenadlung , I found a solution for resetting the height correctly _after_ moving/dragging a node. To be clear, the issue is **not** with `react-virtualized`.
> 
> `react-sortable-tree.js` is missing a call to `List.recomputeRowHeights()`. See docs [here](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#recomputerowheights-index-number) and an example [here](https://plnkr.co/edit/o4ArW2Os4XDdlTV8OuRx?p=preview).
> 
> Right now I have that call within `moveNode` function (which happens after drag is finished) which updates the row heights correctly after a move happens.
> 
> The rowHeights update while dragging is still not working though. I don't know where to call the recomputeRowHeights function to update the rowHeights while dragging. I've tried several different places in the lifecylce, but there's a point between a row of height X and a row of height Y that it sits between "slideRegion" of _both_ rows, so it shifts back and forth from below to above the hovered row. You can see this behavior in the attached gif. I have a debounce in place to slow it down, else it happens so often you can't see it and it nearly crashes. You can see that if I leave that particular "middle ground" slide region and go lower or higher than the row, that it resolves and sets the correct rowHeight both when it's above and below the row with the larger rowHeight.
> 
> ![rowheight issue](https://user-images.githubusercontent.com/5686190/37171828-24b4223e-22dd-11e8-99f3-ff17ec07a17a.gif)
> 
> I can open up a PR on what I have that fixes the rowHeight issue _after_ moving a node, but will need some more time and/or assistance resolving the problem while dragging (detailed above). What do you guys think?


How did you use debounce to solve the problem of moving nodes back and forth?

**Comment by healthyl** [comment]
> i hacked around this by forcing calling recomputeRowHeights on drag change. On drag I set an interval at 250ms to recompute row heights. When stopped dragging, I clear the interval. I originally didnt use an interval and called recomputeRowHeights within the `onChange` prop, but that caused me to call `recomputeRowHeights` hundreds of times (for every pixel i moved my cursor while dragging). Feel free to change 250ms to a lower number. But calling `recomputeRowHeights` too often will hurt performance. This method isnt perfect, but it does the job:
> 
> ```
> <SortableTree>
> ...
> reactVirtualizedListProps={{
>    autoHeight: true,
>    ref: ref => this.list = ref;
> }}
> onDragStateChanged={({ isDragging }) => {
>      if (this.list) {
>        const recompute = () => {
>           this.list.wrappedInstance.recomputeRowHeights();
>         };
>         if (isDragging) {
>           this.dragInterval = setInterval(recompute, 250);
>         } else {
>           clearInterval(this.dragInterval);
>           this.list.wrappedInstance.recomputeRowHeights();
>         }
>     }
> }}
> </SortableTree>
> ```

Hi, thank you very much for a workaround with this code. However i cant get it to run. It always throws an Error saying "TypeError: Cannot set property 'list' of undefined". Cant figure out whats wrong with the ref and the this context. Any help is appreciated.

**Comment by d9k** [comment]
reproduced issue in storybook example at my fork

**Comment by d9k** [comment]
> there's a point between a row of height X and a row of height Y that it sits between "slideRegion" of both rows, so it shifts back and forth from below to above the hovered row

@jcalebj, probably some `dnd-manager.js` patch required, similar to https://github.com/frontend-collective/react-sortable-tree/issues/483#issuecomment-581139473

**Comment by d9k** [comment]
It seems like calling `recomputeRowHeights()` aborts dragging :c

```
  constructor(props) {
  // . . . .
    this.refReactVirtualizedList = React.createRef();

    this.reactVirtualizedListProps = {
      // autoHeight: true,
      ref: this.refReactVirtualizedList,
    };

    this.onDragStateChanged = this.onDragStateChanged.bind(this);
    this.calcRowHeight = this.calcRowHeight.bind(this);
    this.onChange= this.onChange.bind(this);
    this.virtualListRecomputeRowHeights = this.virtualListRecomputeRowHeights.bind(this);
  }

    onDragStateChanged(args) {
      this.virtualListRecomputeRowHeights();
    }

  virtualListRecomputeRowHeights() {
    if (this.props.isVirtualized) {
      // TODO seems like calling recomputeRowHeights() aborts dragging :c
      this.refReactVirtualizedList.current.wrappedInstance.current.recomputeRowHeights();
    }
  }

  render() {
    // . . . . .
    return (
       // . . . . .
       <SortableTree
              // . . . . .
              isVirtualized
              onDragStateChanged={this.onDragStateChanged}
              reactVirtualizedListProps={this.reactVirtualizedListProps}
              rowHeight={this.calcRowHeight}
       />
       // . . . . .
     );
  }

```

this.refReactVirtualizedList.current.wrappedInstance.current.recomputeRowHeights();
```

**Comment by d9k** [comment]
`handleTopDragEndCapture()` at `node_modules/react-dnd-html5-backend/dist/esm/HTML5BackendImpl.js` is called after `recomputeRowHeights()`.

**Comment by d9k** [comment]
> It seems like calling `recomputeRowHeights()` aborts dragging :c

reproduced with an old code (`2019-05-03`) too: https://github.com/d9k/react-sortable-tree/tree/feature-11518-old

**Comment by d9k** [comment]
> ![rowheight issue](https://user-images.githubusercontent.com/5686190/37171828-24b4223e-22dd-11e8-99f3-ff17ec07a17a.gif)

> You can see that if I leave that particular "middle ground" slide region and go lower or higher than the row, that it resolves and sets the correct rowHeight both when it's above and below the row with the larger rowHeight.

Error reproduction with storybook `http://localhost:3001/?path=/story/advanced--row-height-function-react-virtualized` at https://github.com/d9k/react-sortable-tree/tree/feature-11518:

![2023_01_26__14_10_39__1407x1299](https://user-images.githubusercontent.com/7331988/214823603-03f90237-09d6-4e3d-9da6-8bd6a914558b.png)
![2023_01_26__14_10_40__1407x1299](https://user-images.githubusercontent.com/7331988/214823624-ce914eeb-87c6-49e0-8ef6-d641cb2baf9e.png)
![2023_01_26__14_10_46__1407x1299](https://user-images.githubusercontent.com/7331988/214823627-5d1ca4b7-40d9-4f34-9abb-d959389780d9.png)

**Comment by d9k** [comment]
* Added debounced `recomputeRowHeights()` call at the end of `hover()` at `dnd-manager.js`
* Added coordinates check to `hover()` at `dnd-manager.js` to reduce "shifting back and forth from below to above the hovered row" when the mouse is not moved
